File: src/main/java/net/yakclient/mixin/registry/pool/RegistryPool.java
package net.yakclient.mixin.registry.pool;

import net.yakclient.mixin.internal.loader.PackageTarget;

import java.util.*;
import java.util.function.Consumer;

public abstract class RegistryPool<T> {
    protected final Map<Location, PoolQueue<T>> pool;

    public RegistryPool() {
        this.pool = new HashMap<>();
    }

    public abstract Location pool(T type) throws ClassNotFoundException;

    public abstract PackageTarget register(Location location);

    public void empty(Location location) {

        if (this.pool.containsKey(location)) this.register(location);
    }

    public void empty() {
        for (Location location : this.pool.keySet()) {
            this.register(location);
        }
    }

    static class PoolQueue<T> {
        protected final Queue<PoolNode<T>> queue;

        public PoolQueue() {
            this.queue = new LinkedList<>();
        }

        public PoolQueue<T> add(T type, Consumer<PackageTarget> registration) {
            this.queue.add(new PoolNode<>(type, registration));
            return this;
        }

        public PoolQueue<T> add(T type, Consumer<PackageTarget> registration, UUID proxy) {
            this.queue.add(new ProxiedPoolNode<>(type, registration, proxy));
            return this;
        }

        public static class PoolNode<T> {
            private final T value;
            private final Consumer<PackageTarget> registration;

            PoolNode(T value, Consumer<PackageTarget> registration) {
                this.value = value;
                this.registration = registration;
            }

            public T getValue() {
                return this.value;
            }

            public void run(PackageTarget target) {
                this.registration.accept(target);
            }
        }

        public static class ProxiedPoolNode<T> extends PoolNode<T> {
            private final UUID proxy;

            ProxiedPoolNode(T value, Consumer<PackageTarget> registration, UUID proxy) {
                super(value, registration);
                this.proxy = proxy;
            }

            public UUID getProxy() {
                return proxy;
            }
        }
    }
}

File: base/src/main/kotlin/net/yakclient/mixins/base/registry/pool/RegistryPool.kt
package net.yakclient.mixins.base.registry.pool

import java.util.*

abstract class RegistryPool<T> {
    protected val pool: MutableMap<Location, PoolQueue<T>>
    init {
        pool = HashMap()
    }

    @Throws(ClassNotFoundException::class)
   abstract fun pool(type: T): Location

    abstract fun mix(location: Location)

    fun mixAll(location: Location) {
        if (pool.containsKey(location)) mix(location)
    }

    fun mixAll() {
        for (location in pool.keys) {
            mix(location)
        }
    }

    class PoolQueue<T> {
        val queue: Queue<PoolNode<T>>
        init {
            queue = LinkedList()
        }

        fun add(type: T): PoolQueue<T> {
            queue.add(PoolNode(type))
            return this
        }

        fun add(type: T, proxy: UUID): PoolQueue<T> {
            queue.add(ProxiedPoolNode(type, proxy))
            return this
        }

        open class PoolNode<T> internal constructor(val value: T)
        class ProxiedPoolNode<T> internal constructor(value: T, val proxy: UUID) : PoolNode<T>(value)
    }
}